### PR TITLE
Closes #807, Closes #1987, #2446 Switch to CKEditor5 and update text editor and format settings.

### LINKED ISSUE #807: CKEditor 5 Spike
Research and present adding the CKEditor 5 contrib module to QS2.

Start in September 2021

**Current situation**

- CKEditor 4 has EOL same as Drupal 9
- CKEditor 5 is in development - no release yet
- We have a lot of tickets pertaining to improving UX on CKEditor 4
- Migration to CKEditor 5 will need to happen

**Some questions**

- Would moving to CKE5 now impact migration?
- Will CKE5 have a source code feature?

### LINKED ISSUE #1987: Migrate az_standard text format to CKEditor5
## Motivation

Drupal 10 removes CKEditor 4 from core

### Describe the solution you'd like
Update config for az_standard to use CKEditor 5 without data loss.

### Describe alternatives you've considered
Use ckeditor contrib module

### Roles and Permissions considerations
A clear and concise description of how each of the following roles would be impacted by this change:
- Anonymous user
- Authenticated user
- Content editor
- Content administrator
- Administrator

## Additional context

- https://ckeditor.com/docs/ckeditor5/latest/installation/getting-started/migration-from-ckeditor-4.html
- https://www.drupal.org/docs/core-modules-and-themes/core-modules/ckeditor-5-module/plugin-and-contrib-module-development

### PR DESCRIPTION
**Pull request by joeparsons**
## Description

Alternate CKEditor5 PR that _does not_ move `az_barrio`'s Arizona Bootstrap theme settings to `az_core` like #2596 does.

- Upgrades `az_standard` text format and editor configs to CKEditor5
- Upgrades `basic` text format and editor configs to CKEditor5
- Installs CKEditor5 module instead of CKEditor (4) module by default on new site installs
- Adds DB update to install CKEditor5 module on existing sites
- Updates icons for custom embed buttons to SVGs (needed for CKEditor5)
- Adds CKEditor5 stylesheets to AZ Barrio theme
- Moves most of `az_barrio`'s `hook_library_info_alter()` implementation (and associated helper functions) to `az_core` so AZ Barrio libraries (including Arizona Bootstrap) are loaded correctly in admin theme

Questions we still need to consider:
- Is it going to be safe for people to apply these config changes via Config Distro if they have customized the text formats or editors in any way?
- When (if ever) do we want to automate the uninstallation of the contrib ckeditor module?
- Should we remove the remaining CKeditor 4 specific things from Quickstart now or later? (AZ Barrio ckeditor styles, `az_media_ckeditor_styles_alter()`, etc)

See validation messages that came up when switching editor/text_format configs to CKeditor5 here: https://github.com/az-digital/az_quickstart/issues/1987#issuecomment-1640724617

## Related issues

- Closes #807
- Closes #1987
- #2446

## How to test
https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2769.probo.build/
Verify that editor works with new and existing content at least as well as it did with CKEditor 4.

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [x] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_paragraphs/az_paragraphs.libraries.yml [hunk 1]
```diff
@@ -31,16 +29,14 @@ az_paragraphs.az_text_media:
     theme:
       css/az_paragraphs_az_text_media.css: {}
   dependencies:
-    - az_barrio/global-styling
-    - az_barrio/arizona-bootstrap
+    - az_core/arizona-bootstrap
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/az_paragraphs.libraries.yml:32]:
```suggestion
    - az_barrio/arizona-bootstrap
```

File: themes/custom/az_barrio/css/ckeditor5.css [hunk 2]
```diff
@@ -0,0 +1,40 @@
+/* See https://www.drupal.org/project/drupal/issues/3332416 */
+.ck.ck-toolbar.ck-toolbar_grouping > .ck.ck-toolbar__items {
+  flex-wrap: wrap;
+}
+
+/* /*
```
Comment (joeparsons) [inline themes/custom/az_barrio/css/ckeditor5.css:6]:
```suggestion
/*
```

File: themes/custom/az_barrio/css/ckeditor5.css [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * Override the width of the table set by Bootstrap content styles.
+ * See: https://github.com/ckeditor/ckeditor5/issues/3253.
+ */
+ .ck.ck-content .table {
+  width: auto;
+}
+/*
```
Comment (joeparsons) [inline themes/custom/az_barrio/css/ckeditor5.css:13]:
```suggestion
/**
```

**Comment by sanjanans** [comment]
I have completed testing the CK editor. Found very few minor issues. None of them should be a blocker. 

However, you may want to take a look at my test page https://7b5e2985-cedf-4a13-ab91-3a35916c835d--pr-2769.probo.build/sanjana-test. The “text” paragraph type (Table: Testing Embeds) is rendering in a weird way in edit mode. Not sure what caused it. I did try to embed all sorts of stuff in that table. 

Below I’ve listed a few minor issues I discovered:

1. “**Insert code block**” in the CK editor: While adding code, there is no way to click out of it. Of course, there are ways around it, and also there won’t be much need for this feature so this should not be a blocker.
2. In **"Styles": “Dark table header”** does not work, or I do not know how to use it.
3. **Small table** is NOT rendered small on the page though it looks small in edit mode.
4. **Table caption** is vertical in edit mode though it shows properly (horizontal) on the page.
5. Not sure if text alignment and top/bottom margins in “styles” are working. It is possible I do not know the use case for it.

**Comment by danahertzberg** [comment]
I recommend having the same order of buttons to help ease the transition.

File: modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml [hunk 4]
```diff
@@ -4,96 +4,277 @@ dependencies:
...
+      - specialCharacters
+      - '|'
+      - link
+      - '|'
+      - bulletedList
+      - numberedList
+      - '|'
+      - alignment
+      - '|'
+      - insertTable
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:30]:
```suggestion
      - bootstrapGrid
      - insertTable
```

File: modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml [hunk 5]
```diff
@@ -4,96 +4,277 @@ dependencies:
...
+      - drupalMedia
+      - az_embed_content
+      - az_embed_content_block
+      - '|'
+      - sourceEditing
+      - removeFormat
+      - '|'
+      - code
+      - codeBlock
+      - bootstrapGrid
```
Comment (joeparsons) [inline modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:46]:
```suggestion
```

File: themes/custom/az_barrio/az_barrio.info.yml [hunk 6]
```diff
@@ -48,6 +48,10 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
 
+ckeditor5-stylesheets:
+  - css/ckeditor-arizona-bootstrap.min.css
+  - css/ckeditor5.css
```
Comment (joeparsons) [inline themes/custom/az_barrio/az_barrio.info.yml:53]:
```suggestion
  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
  - css/ckeditor5.css
```

File: az_quickstart.install [hunk 7]
```diff
@@ -118,3 +118,10 @@ function az_quickstart_update_920701() {
 function az_quickstart_update_920702() {
   \Drupal::service('module_installer')->install(['az_paragraphs_splitscreen']);
 }
+
+/**
+ * Ensure ckeditor5 module is installed.
+ */
+function az_quickstart_update_920801() {
```
Comment (joeparsons) [inline az_quickstart.install:125]:
```suggestion
function az_quickstart_update_1020801() {
```

File: themes/custom/az_barrio/az_barrio.info.yml [hunk 8]
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
 
+ckeditor5-stylesheets:
+  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
+  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
+  - css/ckeditor5.css
```
Comment (joeparsons) [inline themes/custom/az_barrio/az_barrio.info.yml:54]:
```suggestion
  - css/ckeditor5.css
  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```

File: themes/custom/az_barrio/css/az-barrio-off-canvas-nav.css [hunk 9]
```diff
@@ -6,3 +6,11 @@ header#header_arizona {
...
   border: 0;
 }
+
+/*
+ * Remove when https://github.com/az-digital/arizona-bootstrap/pull/901 is
+ * included in a release
+ */
+.btn-redbar .icon-text {
+  margin: -5px 0;
+}
```
Comment (joeparsons) [inline themes/custom/az_barrio/css/az-barrio-off-canvas-nav.css:16]:
```suggestion
```

File: modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install [hunk 10]
```diff
@@ -5,6 +5,8 @@
  * Install, update and uninstall functions for az_paragraphs_text_media module.
  */
 
+use Drupal\paragraphs\Entity\Paragraph;
```
Comment (trackleft) [inline modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install:8]: Was this important?

Comment (joeparsons) [reply modules/custom/az_paragraphs/az_paragraphs_text_media/az_paragraphs_text_media.install:8]: No.  It was unrelated to the other changes in this PR so I omitted the change.

File: themes/custom/az_barrio/az_barrio.info.yml [hunk 11]
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
 
+ckeditor5-stylesheets:
+  - css/ckeditor5.css
+  - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
+  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```
Comment (trackleft) [inline themes/custom/az_barrio/az_barrio.info.yml:54]:
```suggestion
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest-2.x/css/arizona-bootstrap.min.css
```

File: themes/custom/az_barrio/az_barrio.info.yml [hunk 12]
```diff
@@ -48,6 +48,11 @@ ckeditor_stylesheets:
   - https://cdn.digital.arizona.edu/lib/az-icons/main/az-icons-styles.css
   - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest/css/arizona-bootstrap.min.css
```
Comment (trackleft) [inline themes/custom/az_barrio/az_barrio.info.yml:49]:
```suggestion
  - https://cdn.digital.arizona.edu/lib/arizona-bootstrap/latest-2.x/css/arizona-bootstrap.min.css
```

File: modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml [hunk 13]
```diff
@@ -4,96 +4,277 @@ dependencies:
...
+        - xs
+        - sm
+        - md
+        - lg
+        - xl
+    linkit_extension:
+      linkit_enabled: true
+      linkit_profile: az_linkit
+    media_media:
+      allow_view_mode_override: true
```
Comment (trackleft) [inline modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:277]: What is this `allow_view_mode_override` setting?

File: themes/custom/az_barrio/az_barrio.info.yml (hunk 11 above)
Comment (joeparsons) [reply themes/custom/az_barrio/az_barrio.info.yml:54]: Thanks

File: modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml (hunk 13 above)
Comment (joeparsons) [reply modules/custom/az_paragraphs/config/optional/editor.editor.az_standard.yml:277]: Not sure.  Was automatically added by the editor config upgrade.

**Comment by joeparsons** [comment]
Updates since 2023-09-29 Friday Meeting:

- Updated patches for Embed and Entity Embed
- Reordered buttons in AZ Standard editor config to match CKeditor 4
- Replaced experimental build of AZ Bootstrap with latest 2.x version from CDN (basically matching CKEditor 4 functionality for now)
- Removed `hook_library_info_alter()` changes since they don't actually seem needed until we're ready to properly solve #2598 
- Removed a couple of unrelated changes

**Review by trackleft — APPROVED** [review]

**Comment by danahertzberg** [comment]
Testing as a content editor and the functionality is looking great. Improved UX yet still similar to what was there before.

https://53ebdf30-2d20-43df-bbe3-c19a2c93ce72.probo.build/dana-content-editor

**Review by tadean — APPROVED** [review]

**Comment by danahertzberg** [comment]
Not blockers imo. But there are some styles that don't render properly in the styles dropdown/editing field.
- Triangle bullets
- Button outline white (needs background color)
- Text white (needs background color)
- Sanserif text not needed anymore
- Text align top
- Text align middle
- Text align bottom
- Zero margin top
- Zero margin bottom
- Table styles?

**Review by danahertzberg — APPROVED** [review]

**Comment by sanjanans** [comment]
There seems to be a slight issue with the bootstrap grid in edit mode. On [Sanjana Test page](https://53ebdf30-2d20-43df-bbe3-c19a2c93ce72.probo.build/sanjana-test) I created a 3 col grid and inserted a video in the 1st column. It took up the space for the 2nd col as well. I had to click inside the 1st grid to add text in the 2nd col.
This issue is only in the admin interface. Once saved, it renders correctly on the page. 
Screenshot 2023-10-03 at 12 15 08 PM

**Review by sanjanans — APPROVED** [review]
This is ready!! Thumbs up from me.

**Pull request merged**